File: challenges-in-python/two_heaps/maximize_capital.py
# Solving the "Maximize Capital" Challenge
# Challenge Description:
# Given two arrays: Capital and Profits of equal length, and an integer k. 
# Each ith element of Capital denotes the capital you need to start the ith project, 
# and each ith element of Profits denotes the profit you can earn from completing the ith project. 
# Starting with 0 capital, you can only start a project if you have the required capital. 
# Find the maximum capital you can accumulate after finishing at most k projects. 
# You can only work on one project at a time.

# Approach:
# Use a min-heap to keep track of available projects based on capital requirements.
# Use a max-heap to keep track of profitable projects that can be started with the available capital.

import heapq

def find_maximum_capital(capital, profits, k, initial_capital):
    min_capital_heap = [(capital[i], profits[i]) for i in range(len(capital))]
    max_profit_heap = []
    
    # Build the min-heap based on capital requirements
    heapq.heapify(min_capital_heap)

    for _ in range(k):
        #  Move all projects that can be started with the aviailable capital to the max-heap
        while min_capital_heap and min_capital_heap[0][0] <= initial_capital:
            capital, profit = heapq.heappop(min_capital_heap)
            heapq.heappush(max_profit_heap, (-profit, capital)) # Use -profit to make it a max-heap, the negative sign is removed when the profit is popped
        
        # Break if no project can be started
        if not max_profit_heap:
            break
        # Start the most profitable project
        initial_capital += -heapq.heappop(max_profit_heap)[0] # Remove the negative sign to get the profit

    return initial_capital

# Test
def test_find_maximum_capital():
    capital = [0, 1, 2]
    profits = [1, 2, 3]
    k = 2
    initial_capital = 1
    assert find_maximum_capital(capital, profits, k, initial_capital) == 6
    print("All test cases passed")

test_find_maximum_capital()

File: challenges-in-rust/src/two_heaps/maximize_capital.rs
// Solving the "Maximize Capital" Challenge
// Problem Statement
// Given two arrays: Capital and Profits of equal length and an integer k. 
// Each element in Capital shows the capital that must be invested to start the corresponding project in Profits. 
// You are given an initial amount of capital initial_capital. You can only start at most k projects. 
// You can only start a project if you have enough capital. 
// Once a project is started, you can assume you instantly gain its profit which can then be used to start other projects. 
// Find the maximum capital you can achieve after finishing at most k projects.

// Implementation Strategy
// Use two heaps (priority queues in Rust):

// A min-heap to keep track of projects available within the current capital (W).
// A max-heap to keep track of profitable projects that can be started.
// Rust Implementation
// Assuming Rust's standard library doesn't have a dedicated max-heap, 
// we use BinaryHeap with reversed profit values for the min-heap and tuples of (Profit, Capital) for sorting.

use std::collections::BinaryHeap;

pub fn find_maximum_capital(capital: Vec<i32>, profits: Vec<i32>, initial_capital: i32, k: i32) -> i32 {
    println!("capital: {:?}", capital);
    println!("profits: {:?}", profits);
    println!("initial_capital: {:?}", initial_capital);
    println!("k: {:?}", k);
    let mut available_capital = initial_capital;
    let mut projects = capital.into_iter().zip(profits).collect::<Vec<_>>();
    println!("projects: {:?}", projects);
    projects.sort_by_key(|x| x.0); // Sort by capital required in ascending order
    let mut available_projects = BinaryHeap::new();

    for _ in 0..k {
        while let Some(&next_project) = projects.iter().find(|&&(capital, _)| capital <= available_capital) { // Find the next project that can be started
            available_projects.push(std::cmp::Reverse(next_project.1)); // Push the profit value to a max-heap
            projects.retain(|&x| x != next_project); // Remove the project from the list
            println!("available_projects: {:?}", available_projects);
            println!("projects: {:?}", projects);
            println!("next_project: {:?}", next_project);
            println!("available_capital: {:?}", available_capital);
            println!("k: {:?}", k);
        }
        if let Some(std::cmp::Reverse(profit)) = available_projects.pop() { // Pop the max profit value
            available_capital += profit;
            println!("profit: {:?}", profit);
        } else {
            break; // No projects can be started
        }
    }

    available_capital
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_maximum_capital() {
        assert_eq!(find_maximum_capital(vec![0, 1, 2], vec![1, 2, 3], 1, 2), 4);
        assert_eq!(find_maximum_capital(vec![0, 1, 2, 3], vec![1, 2, 3, 5], 0, 3), 6);
    }
}